### PR TITLE
cylc gsummary/gscan, gpanel: use new cylc scan features

### LINKED ISSUE #1488: Gsummary should use the new port scanning functionality.
From #1480  ```cylc scan``` can retrieve task state totals directly from all suite daemons on all scanned hosts, without requiring authentication.  ```cylc gsummary``` needs to be updated to use the same information.  Currently it gets task state totals from ```cylc cat-state```, which requires file access rights (we will still need this for suite stop summaries, however).

@benfitzpatrick - as the architect of gsummary, can you knock this one off?  It should be a fairly extensive simplification, as ```cylc scan``` now prints out exactly the required information.

### PR DESCRIPTION
**Pull request by benfitzpatrick**
This closes #1488.

@hjoliver, please review 1.

### CODE REVIEW

**Comment by hjoliver** [comment]
```

$ Traceback (most recent call last):
  File "/home/oliverh/cylc/cylc.git/lib/cylc/gui/gscan.py", line 651, in _on_query_tooltip
    tooltip.set_text(text)
UnboundLocalError: local variable 'text' referenced before assignment
```

**Comment by hjoliver** [comment]
(I was just going to say - that's caused by purely runahead cycles :grin:)

**Comment by benfitzpatrick** [comment]
I guessed so!

**Comment by hjoliver** [comment]
From a quick look, this doesn't yet use all of the new ```scan``` functionality - e.g. display other users on the same host, or filtering by suite name and owner  (c.f. ```cylc scan``` options ```--all```, ```--name```, ```--owner```)?    Do you want to leave that to a follow-up change?

**Comment by benfitzpatrick** [comment]
Fair point! I hadn't thought of that. Could we have that in a follow-up?

**Comment by hjoliver** [comment]
OK, a follow-up will do.

**Comment by hjoliver** [comment]
Review 1: good, can't see any other problems.

**Comment by benfitzpatrick** [comment]
@kaday, please review 2. This goes over some of the ground from your update time change, since the update time is now contained in the scan results.

File: lib/cylc/gui/gscan.py [hunk 1]
```diff
@@ -64,20 +91,25 @@ def get_host_suites(hosts, timeout=None, owner=None):
...
-        for line in popen.communicate()[0].splitlines():
-            if line:
-                name, _, host, _ = line.split()
-                if host not in host_suites_map:
-                    host_suites_map[host] = []
-                host_suites_map[host].append(name)
-        
+        host_suites_map = parse_cylc_scan_raw(popen.communicate()[0])
+    else:
+        print >>sys.stderr, popen.communicate()[1]
```
Comment (kaday) [inline lib/cylc/gui/gscan.py]: Operator not followed by a space.

File: lib/cylc/gui/gscan.py [hunk 2]
```diff
@@ -89,32 +121,35 @@ def get_task_cycle_statuses_updatetime(host, suite, owner=None):
...
-    suite_update_times = re.search("^time : [^ ]+ \(([0-9]+)\)$", out, re.M)
-    if suite_update_times is not None:
-        suite_update_times = int(suite_update_times.group(1))
+        task, state = task_result.groups()
+        task_name, task_point = task.split(".")
+        for states_point in ("states", "states:" + task_point):
+            suite_info.setdefault(states_point, {})
+            suite_info[states_point].setdefault(state, 0)
+            suite_info[states_point][state] += 1
+    suite_update_time_match = re.search("^time : [^ ]+ \(([0-9]+)\)$", out, re.M)
```
Comment (kaday) [inline lib/cylc/gui/gscan.py]: Line too long

File: lib/cylc/gui/gscan.py [hunk 3]
```diff
@@ -739,30 +772,29 @@ def run(self):
...
-            for host in statuses:
-                for suite in statuses[host]:
-                    prev_suites.append((host, suite))
-            self.prev_suites = prev_suites
-            self.statuses = statuses
-            self.stop_summaries = stop_summaries
+            self.hosts_suites_info, self.stopped_hosts_suites_info = (
+                update_hosts_suites_info(
+                   self.hosts, self.owner,
+                   prev_stopped_hosts_suites_info=self.stopped_hosts_suites_info,
```
Comment (kaday) [inline lib/cylc/gui/gscan.py]: Line too long, but not sure you can do anything about it.

File: lib/cylc/gui/gscan.py [hunk 4]
```diff
@@ -911,134 +939,103 @@ def _get_user_expanded_row_ids(self):
...
+                                                key=lambda _: _[1]):
+                        if state != "runahead":
+                            # 'runahead' states are usually hidden.
+                            states_text += '%s %d ' % (state, number)
+                    if not states_text:
+                        # Purely runahead cycle.
+                        continue
+                    model_data.append(states_text.rstrip())
+                    if key == "states":
+                        parent_iter = self.suite_treemodel.append(None, model_data)
```
Comment (kaday) [inline lib/cylc/gui/gscan.py]: Line too long

File: lib/cylc/gui/gscan.py [hunk 5]
```diff
@@ -911,134 +939,103 @@ def _get_user_expanded_row_ids(self):
...
-
-
-def get_new_statuses_and_stop_summaries_updatetime(hosts, owner,
-                                        prev_stop_summaries=None,
-                                        prev_suites=None,
-                                        stop_suite_clear_time=86400):
-    """Return dictionaries of statuses, stop_summaries and updatetimes."""
+
+def update_hosts_suites_info(hosts, owner, prev_stopped_hosts_suites_info=None,
+                             prev_hosts_suites=None, stop_suite_clear_time=86400):
```
Comment (kaday) [inline lib/cylc/gui/gscan.py]: Line too long

**Comment by kaday** [comment]
In gscan with the updated column showing, the tooltip reports that the info was retrieved at the time of the last suite update. This should be corrected to the time at which gscan retrieved this information.

**Comment by benfitzpatrick** [comment]
> In gscan with the updated column showing, the tooltip reports that the info was retrieved at the time of the last suite update. This should be corrected to the time at which gscan retrieved this information.

This is now (for new-scan suites) the suite's own state-change time, although of course it can be out of date... I've removed the descriptive text and kept the date-time.

**Comment by kaday** [comment]
Make the hoover-over the polled time please as discussed with Andy and I.

**Comment by benfitzpatrick** [comment]
Now updated with the polled time vs the last-state-updated time in the hover-over

**Comment by benfitzpatrick** [comment]
(and rebased)

**Comment by kaday** [comment]
As discussed with Ben, this needs fixing as if inactive for ~20 mins then a trace-back is printed to the terminal.

**Comment by benfitzpatrick** [comment]
There is some kind of condition that leads to `cylc scan` aborting nastily - although this isn't technically an issue with the code change here, it needs to be looked into.

**Comment by matthewrmshin** [comment]
@benfitzpatrick This change in my branch of the same name should fix various nasty trace back problems f868542655c1cca8a833b309d6a1590744d53f1c - please consider merging it into your branch.

**Comment by benfitzpatrick** [comment]
Merged in Matt's changes. Please re-review with f868542.

**Comment by kaday** [comment]
Works full now.

**Pull request merged**